nconnect.go

```diff
@@ -245,234 +244,271 @@ func Run(opts *config.Opts) {
+	log.Println("Client socks proxy listen address:", nc.opts.LocalSocksAddr)
 
-		tun, err = tunnel.NewTunnel(account, opts.Identifier, ssAddr, remoteTunnelAddr, opts.Tuna, tunnelConfig)
+	if nc.opts.Tun || nc.opts.VPN {
+		tunDevice, err := arch.OpenTunDevice(nc.opts.TunName, nc.opts.TunAddr, nc.opts.TunGateway, nc.opts.TunMask, nc.opts.TunDNS, true)
 		if err != nil {
-			log.Fatal(err)
+			return fmt.Errorf("failed to open TUN device: %v", err)
 		}
 
-		log.Println("Client NKN address:", tun.Addr().String())
-		log.Println("Client socks proxy listen address:", opts.LocalSocksAddr)
+		core.RegisterOutputFn(tunDevice.Write)
 
-		if opts.Tun || opts.VPN {
-			tunDevice, err := arch.OpenTunDevice(opts.TunName, opts.TunAddr, opts.TunGateway, opts.TunMask, opts.TunDNS, true)
-			if err != nil {
-				log.Fatalf("Failed to open TUN device: %v", err)
-			}
+		core.RegisterTCPConnHandler(socks.NewTCPHandler(proxyHost, proxyPort))
+		core.RegisterUDPConnHandler(socks.NewUDPHandler(proxyHost, proxyPort, 30*time.Second))
 
-			core.RegisterOutputFn(tunDevice.Write)
+		lwipWriter := core.NewLWIPStack()
 
-			core.RegisterTCPConnHandler(socks.NewTCPHandler(proxyHost, proxyPort))
-			core.RegisterUDPConnHandler(socks.NewUDPHandler(proxyHost, proxyPort, 30*time.Second))
+		go func() {
+			_, err := io.CopyBuffer(lwipWriter, tunDevice, make([]byte, mtu))
+			if err != nil {
+				log.Fatalf("Failed to write data to network stack: %v", err)
+			}
+		}()
 
-			lwipWriter := core.NewLWIPStack()
+		log.Println("Started tun2socks")
 
-			go func() {
-				_, err := io.CopyBuffer(lwipWriter, tunDevice, make([]byte, mtu))
+		if nc.opts.VPN {
+			for _, dest := range vpnCIDR {
+				log.Printf("Adding route %s", dest)
+				out, err := arch.AddRouteCmd(dest, nc.opts.TunGateway, nc.opts.TunName)
+				if len(out) > 0 {
+					os.Stdout.Write(out)
+				}
 				if err != nil {
-					log.Fatalf("Failed to write data to network stack: %v", err)
+					os.Stdout.Write([]byte(util.ParseExecError(err)))
+					os.Exit(1)
 				}
-			}()
-
-			log.Println("Started tun2socks")
-
-			if opts.VPN {
-				for _, dest := range vpnCIDR {
-					log.Printf("Adding route %s", dest)
-					out, err := arch.AddRouteCmd(dest, opts.TunGateway, opts.TunName)
+				defer func(dest *net.IPNet) {
+					log.Printf("Deleting route %s", dest)
+					out, err := arch.DeleteRouteCmd(dest, nc.opts.TunGateway, nc.opts.TunName)
 					if len(out) > 0 {
 						os.Stdout.Write(out)
 					}
 					if err != nil {
 						os.Stdout.Write([]byte(util.ParseExecError(err)))
-						os.Exit(1)
 					}
-					defer func(dest *net.IPNet) {
-						log.Printf("Deleting route %s", dest)
-						out, err := arch.DeleteRouteCmd(dest, opts.TunGateway, opts.TunName)
-						if len(out) > 0 {
-							os.Stdout.Write(out)
-						}
-						if err != nil {
-							os.Stdout.Write([]byte(util.ParseExecError(err)))
-						}
-					}(dest)
-				}
+				}(dest)
 			}
 		}
 	}
 
-	if opts.Server {
-		err = (&opts.Config).VerifyServer()
+	go func() {
```

Comment:
These two goroutines are duplicated in both StartClient and StartServer functions

Reply:
Yes. moved into a method.